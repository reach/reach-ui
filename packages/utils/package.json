{
  "name": "@reach/utils",
  "version": "0.8.0",
  "description": "Internal, shared utilities for Reach UI.",
  "author": "React Training <hello@reacttraining.com>",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/reach/reach-ui.git",
    "directory": "packages/utils"
  },
  "scripts": {
    "build": "yarn clean && yarn compile",
    "clean": "rm -rf ./dist",
    "compile": "cross-env NODE_ENV=production tsdx build --format=cjs,esm --tsconfig tsconfig.build.json"
  },
  "peerDependencies": {
    "react": "^16.8.0",
    "react-dom": "^16.8.0"
  },
  "main": "dist/index.js",
  "module": "dist/utils.esm.js",
  "typings": "dist/index.d.ts",
  "files": [
    "README.md",
    "dist"
  ],
  "dependencies": {
    "tslib": "1.10.0"
  }
}
