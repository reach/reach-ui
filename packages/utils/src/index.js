import { useRef, useCallback, useMemo, useEffect } from "react";

let checkedPkgs = {};

let checkStyles = () => {};

if (__DEV__) {
  checkStyles = pkg => {
    // only check once per package
    if (checkedPkgs[pkg]) return;
    checkedPkgs[pkg] = true;

    if (
      parseInt(
        window
          .getComputedStyle(document.body)
          .getPropertyValue(`--reach-${pkg}`),
        10
      ) !== 1
    ) {
      console.warn(
        `@reach/${pkg} styles not found. If you are using a bundler like webpack or parcel include this in the entry file of your app before any of your own styles:

    import "@reach/${pkg}/styles.css";

  Otherwise you'll need to include them some other way:

    <link rel="stylesheet" type="text/css" href="node_modules/@reach/${pkg}/styles.css" />

  For more information visit https://ui.reach.tech/styling.
  `
      );
    }
  };
}

export { checkStyles };

export let wrapEvent = (theirHandler, ourHandler) => event => {
  theirHandler && theirHandler(event);
  if (!event.defaultPrevented) {
    return ourHandler(event);
  }
};

export const assignRef = (ref, value) => {
  if (ref == null) return;
  if (typeof ref === "function") {
    ref(value);
  } else {
    try {
      ref.current = value;
    } catch (error) {
      throw new Error(`Cannot assign value "${value}" to ref "${ref}"`);
    }
  }
};

export function useUpdateEffect(effect, deps) {
  const mounted = useRef(false);
  useEffect(() => {
    if (mounted.current) {
      effect();
    } else {
      mounted.current = true;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}

export function useForkedRef(...refs) {
  return useMemo(() => {
    if (refs.every(ref => ref == null)) {
      return null;
    }
    return node => {
      refs.forEach(ref => {
        assignRef(ref, node);
      });
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, refs);
}
