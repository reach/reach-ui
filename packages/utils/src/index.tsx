/* eslint-disable no-unused-vars */

import React, {
  cloneElement,
  createContext,
  isValidElement,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useRef,
  useState
} from "react";
import {
  As,
  AssignableRef,
  ComponentWithAs,
  ComponentWithForwardedRef,
  DistributiveOmit,
  PropsFromAs,
  PropsWithAs
} from "./types";

/**
 * React currently throws a warning when using useLayoutEffect on the server.
 * To get around it, we can conditionally useEffect on the server (no-op) and
 * useLayoutEffect in the browser. We occasionally need useLayoutEffect to
 * ensure we don't get a render flash for certain operations, but we may also
 * need affected components to render on the server. One example is when setting
 * a component's descendants to retrieve their index values.
 *
 * Important to note that using this hook as an escape hatch will break the
 * eslint dependency warnings unless you rename the import to `useLayoutEffect`.
 * Use sparingly only when the effect won't effect the rendered HTML to avoid
 * any server/client mismatch.
 *
 * If a useLayoutEffect is needed and the result would create a mismatch, it's
 * likely that the component in question shouldn't be rendered on the server at
 * all, so a better approach would be to lazily render those in a parent
 * component after client-side hydration.
 *
 * TODO: We are calling useLayoutEffect in a couple of places that will likely
 * cause some issues for SSR users, whether the warning shows or not. Audit and
 * fix these.
 *
 * https://gist.github.com/gaearon/e7d97cdf38a2907924ea12e4ebdf3c85
 * https://github.com/reduxjs/react-redux/blob/master/src/utils/useIsomorphicLayoutEffect.js
 *
 * @param effect
 * @param deps
 */
export const useIsomorphicLayoutEffect = canUseDOM()
  ? React.useLayoutEffect
  : React.useEffect;

const useLayoutEffect = useIsomorphicLayoutEffect;

let checkedPkgs: { [key: string]: boolean } = {};

/**
 * When in dev mode, checks that styles for a given @reach package are loaded.
 *
 * @param packageName Name of the package to check.
 * @example checkStyles("dialog") will check for styles for @reach/dialog
 */
// @ts-ignore
let checkStyles = (packageName: string): void => {};

if (__DEV__) {
  checkStyles = (pkg: string) => {
    // only check once per package
    if (checkedPkgs[pkg]) return;
    checkedPkgs[pkg] = true;

    if (
      process.env.NODE_ENV !== "test" &&
      parseInt(
        window
          .getComputedStyle(document.body)
          .getPropertyValue(`--reach-${pkg}`),
        10
      ) !== 1
    ) {
      console.warn(
        `@reach/${pkg} styles not found. If you are using a bundler like webpack or parcel include this in the entry file of your app before any of your own styles:

    import "@reach/${pkg}/styles.css";

  Otherwise you'll need to include them some other way:

    <link rel="stylesheet" type="text/css" href="node_modules/@reach/${pkg}/styles.css" />

  For more information visit https://ui.reach.tech/styling.
  `
      );
    }
  };
}

export { checkStyles };

/**
 * Passes or assigns an arbitrary value to a ref function or object.
 *
 * @param ref
 * @param value
 */
export function assignRef<T = any>(
  ref: AssignableRef<T> | undefined,
  value: any
) {
  if (ref == null) return;
  if (typeof ref === "function") {
    ref(value);
  } else {
    try {
      // @ts-ignore
      ref.current = value;
    } catch (error) {
      throw new Error(`Cannot assign value "${value}" to ref "${ref}"`);
    }
  }
}

export function boolOrBoolString(value: any) {
  return value === "false" ? false : Boolean(value);
}

export function canUseDOM() {
  return (
    typeof window !== "undefined" &&
    typeof window.document !== "undefined" &&
    typeof window.document.createElement !== "undefined"
  );
}

export function cloneValidElement<P>(
  element: React.ReactElement<P> | React.ReactNode,
  props?: Partial<P> & React.Attributes,
  ...children: React.ReactNode[]
): React.ReactElement<P> | React.ReactNode {
  if (!isValidElement(element)) {
    return element;
  }
  return cloneElement(element, props, ...children);
}

export function createNamedContext<T>(
  name: string,
  defaultValue: T
): React.Context<T> {
  const Ctx = createContext<T>(defaultValue);
  Ctx.displayName = name;
  return Ctx;
}

export function findLastIndex<T = any>(
  array: T[],
  predicate: (element: T, index?: number, arr?: T[]) => boolean
): number {
  let length = array.length >>> 0;
  if (!length) {
    return -1;
  }
  let n = length - 1;
  while (n >= 0) {
    let value = array[n];
    if (predicate(value, n, array)) {
      return n;
    }
    --n;
  }
  return -1;
}

/**
 * Get the scrollbar offset distance.
 */
export function getScrollbarOffset() {
  try {
    if (window.innerWidth > document.documentElement.clientWidth) {
      return window.innerWidth - document.documentElement.clientWidth;
    }
  } catch (err) {}
  return 0;
}

export function isUndefined(value: any) {
  return typeof value === "undefined";
}

/**
 * Joins strings to format IDs for compound components.
 *
 * @param args
 */
export function makeId(...args: (string | number | null | undefined)[]) {
  return args.filter(val => val != null).join("--");
}

/**
 * No-op function.
 */
export function noop(): void {}

// React hook for creating a value exactly once.
// https://github.com/Andarist/use-constant
export function useConstant<T>(fn: () => T): T {
  const ref = React.useRef<{ v: T }>();

  if (!ref.current) {
    ref.current = { v: fn() };
  }

  return ref.current.v;
}

/**
 * Passes or assigns a value to multiple refs (typically a DOM node). Useful for
 * dealing with components that need an explicit ref for DOM calculations but
 * also forwards refs assigned by an app.
 *
 * @param refs Refs to fork
 */
export function useForkedRef<T = any>(
  ...refs: (AssignableRef<T> | undefined)[]
) {
  return useMemo(() => {
    if (refs.every(ref => ref == null)) {
      return null;
    }
    return (node: any) => {
      refs.forEach(ref => {
        assignRef(ref, node);
      });
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, refs);
}

/**
 * Returns the previous value of a reference after a component update.
 *
 * @param value
 */
export function usePrevious<T = any>(value: T) {
  const ref = useRef<T | null>(null);
  useEffect(() => {
    ref.current = value;
  }, [value]);
  return ref.current;
}

/**
 * Call an effect after a component update, skipping the initial mount.
 *
 * @param effect Effect to call
 * @param deps Effect dependency list
 */
export function useUpdateEffect(effect: () => any, deps?: any[]) {
  const mounted = useRef(false);
  useEffect(() => {
    if (mounted.current) {
      effect();
    } else {
      mounted.current = true;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}

/**
 * Wraps a lib-defined event handler and a user-defined event handler, returning
 * a single handler that allows a user to prevent lib-defined handlers from
 * firing.
 *
 * @param theirHandler User-supplied event handler
 * @param ourHandler Library-supplied event handler
 */
export function wrapEvent<E extends React.SyntheticEvent | Event>(
  theirHandler: ((event: E) => any) | undefined,
  ourHandler: (event: E) => any
): (event: E) => any {
  return event => {
    theirHandler && theirHandler(event);
    if (!event.defaultPrevented) {
      return ourHandler(event);
    }
  };
}

/**
 * This is a hack for sure. The thing is, getting a component to intelligently
 * infer props based on a component or JSX string passed into an `as` prop is
 * kind of a huge pain. Getting it to work and satisfy the constraints of
 * `forwardRef` seems dang near impossible. To avoid needing to do this awkward
 * type song-and-dance every time we want to forward a ref into a component
 * that accepts an `as` prop, we abstract all of that mess to this function for
 * the time time being.
 *
 * TODO: Eventually we should probably just try to get the type defs above
 * working across the board, but ain't nobody got time for that mess!
 *
 * @param Comp
 */
export function forwardRefWithAs<P, T extends As>(
  comp: (props: PropsFromAs<T, P>, ref: React.RefObject<any>) => JSX.Element
) {
  return React.forwardRef(comp as any) as ComponentWithAs<T, P>;
}

// Export types
export {
  As,
  AssignableRef,
  ComponentWithAs,
  ComponentWithForwardedRef,
  DistributiveOmit,
  PropsFromAs,
  PropsWithAs
};

////////////////////////////////////////////////////////////////////////////////
// TODO: Move to @reach/descendants once fully tested and implemented

export type Descendant<T, P = {}> = P & {
  element: (T extends HTMLElement ? T : HTMLElement) | null;
};

export interface IDescendantContext<T, P> {
  descendants: Descendant<T, P>[];
  registerDescendant(descendant: Descendant<T, P>): void;
  unregisterDescendant(
    element: (T extends HTMLElement ? T : HTMLElement) | null
  ): void;
}

export function createDescendantContext<T, P = {}>(
  name: string,
  initialValue = {}
) {
  return createNamedContext(name, {
    descendants: [],
    registerDescendant: noop,
    unregisterDescendant: noop,
    ...initialValue
  } as IDescendantContext<T, P>);
}

/**
 * This hook registers our descendant by passing it into an array. We can then
 * search that array by to find its index when registering it in the component.
 * We use this for focus management, keyboard navigation, and typeahead
 * functionality for some components.
 *
 * The hook accepts the element node and (optionally) a key. The key is useful
 * if multiple descendants have identical text values and we need to
 * differentiate siblings for some reason.
 *
 * Our main goals with this are:
 *   1) maximum composability,
 *   2) minimal API friction
 *   3) SSR compatibility*
 *   4) concurrent safe
 *   5) index always up-to-date with the tree despite changes
 *   6) works with memoization of any component in the tree (hopefully)
 *
 * * As for SSR, the good news is that we don't actually need the index on the
 * server for most use-cases, as we are only using it to determine the order of
 * composed descendants for keyboard navigation. However, in the few cases where
 * this is not the case, we can require an explicit index from the app.
 */
export function useDescendant<T, P>(
  {
    context,
    element,
    ...rest
  }: Descendant<T, P> & { context: React.Context<IDescendantContext<T, P>> },
  indexProp?: number
) {
  let [, forceUpdate] = useState();
  let { registerDescendant, unregisterDescendant, descendants } = useContext(
    context
  );

  // Prevent any flashing
  useLayoutEffect(() => {
    if (!element) forceUpdate({});
    // @ts-ignore
    registerDescendant({ element, ...rest });
    return () => unregisterDescendant(element);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [element, ...Object.values(rest)]);

  return (
    indexProp ?? descendants.findIndex(({ element: _el }) => _el === element)
  );
}

export function useDescendants<T, P = {}>() {
  return useState<Descendant<T, P>[]>([]);
}

export function DescendantProvider<T, P>({
  context: Ctx,
  children,
  items,
  set
}: {
  context: React.Context<IDescendantContext<T, P>>;
  children: React.ReactNode;
  items: Descendant<T, P>[];
  set: React.Dispatch<React.SetStateAction<Descendant<T, P>[]>>;
}) {
  let registerDescendant = React.useCallback(
    ({ element, ...rest }: Descendant<T, P>) => {
      if (!element) {
        return;
      }

      set(items => {
        if (items.find(({ element: _el }) => _el === element) == null) {
          /*
           * When registering a descendant, we need to make sure we insert in
           * into the array in the same order that it appears in the DOM. So as
           * new descendants are added or maybe some are removed, we always know
           * that the array is up-to-date and correct.
           *
           * So here we look at our registered descendants and see if the new
           * element we are adding appears earlier than an existing descendant's
           * DOM node via `node.compareDocumentPosition`. If it does, we insert
           * the new element at this index. Because `registerDescendant` will be
           * called in an effect every time the descendants state value changes,
           * we should be sure that this index is accurate when descendent
           * elements come or go from our component.
           */
          let index = items.findIndex(({ element: existingElement }) => {
            if (!existingElement || !element) {
              return false;
            }
            /*
             * Does this element's DOM node appear before another item in the
             * array in our DOM tree? If so, return true to grab the index at
             * this point in the array so we know where to insert the new
             * element.
             */
            return Boolean(
              existingElement.compareDocumentPosition(element) &
                Node.DOCUMENT_POSITION_PRECEDING
            );
          });

          let newItem = { element, ...rest } as Descendant<T, P>;

          // If an index is not found we will push the element to the end.
          if (index === -1) {
            return [...items, newItem];
          }
          return [...items.slice(0, index), newItem, ...items.slice(index)];
        }
        return items;
      });
    },
    /*
     * setDescendants is a state setter initialized by the useDescendants hook.
     * We can safely ignore the lint warning here because it will not change
     * between renders.
     */
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  let unregisterDescendant = useCallback(
    (element: Descendant<T>["element"]) => {
      if (!element) {
        return;
      }

      set(items => items.filter(({ element: _el }) => element !== _el));
    },
    /*
     * setDescendants is a state setter initialized by the useDescendants hook.
     * We can safely ignore the lint warning here because it will not change
     * between renders.
     */
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []
  );

  // Not sure about this just yet, may bail on this and let components deal
  /* let focusNodes = descendants
    .filter(({ disabled }) => !disabled)
    .map(({ element }) => element); */

  // @ts-ignore
  const value: IDescendantContext<T, P> = useMemo(() => {
    return {
      descendants: items,
      registerDescendant,
      unregisterDescendant
    };
  }, [items, registerDescendant, unregisterDescendant]);

  return <Ctx.Provider value={value}>{children}</Ctx.Provider>;
}
