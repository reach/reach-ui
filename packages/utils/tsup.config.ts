import type { defineConfig } from "tsup";
import { getTsupConfig, getPackageInfo } from "@reach-internal/dev/tsup.js";

type TsupConfig = ReturnType<typeof defineConfig>;

let { name: packageName, version: packageVersion } = getPackageInfo(__dirname);
let cfg: TsupConfig = getTsupConfig("src/index.ts", {
  packageName,
  packageVersion,
});
export default cfg;
