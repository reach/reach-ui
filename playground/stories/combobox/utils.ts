import * as React from "react";
import { matchSorter } from "match-sorter";
import cities from "./cities";
import type { City } from "./cities";

export function useCityMatch(term: string): City[] | null {
  let throttledTerm = useThrottle(term, 100);
  return React.useMemo(
    () =>
      term.trim() === ""
        ? null
        : matchSorter(cities, term, {
            keys: [(item) => `${item.city}, ${item.state}`],
          }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [throttledTerm]
  );
}

export function useThrottle(value: any, limit: number) {
  const [throttledValue, setThrottledValue] = React.useState(value);
  const lastRan = React.useRef(Date.now());

  React.useEffect(() => {
    const handler = window.setTimeout(() => {
      if (Date.now() - lastRan.current >= limit) {
        setThrottledValue(value);
        lastRan.current = Date.now();
      }
    }, limit - (Date.now() - lastRan.current));

    return () => window.clearTimeout(handler);
  }, [value, limit]);

  return throttledValue;
}
