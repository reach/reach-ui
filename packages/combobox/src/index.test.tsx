import React, { useState, useMemo } from "react";
import { render } from "$test/utils";
import userEvent from "@testing-library/user-event";
import {
  Combobox,
  ComboboxInput,
  ComboboxList,
  ComboboxOption,
  ComboboxPopover,
} from "@reach/combobox";
import matchSorter from "match-sorter";
import { useThrottle } from "../examples/use-throttle";
import cities from "../examples/cities";

describe("<Combobox />", () => {
  it("should match the snapshot", () => {
    let { asFragment, getByTestId, getByRole } = render(<BasicCombobox />);
    let input = getByTestId("input");
    expect(asFragment()).toMatchSnapshot();
    userEvent.type(input, "e");
    expect(asFragment()).toMatchSnapshot();
    expect(getByRole("combobox")).toHaveAttribute("aria-expanded", "true");
  });
});

////////////////////////////////////////////////////////////////////////////////
function BasicCombobox() {
  let [term, setTerm] = useState("");
  let results = useCityMatch(term);

  const handleChange = (event: any) => {
    setTerm(event.target.value);
  };

  return (
    <div>
      <h2>Clientside Search</h2>
      <Combobox id="holy-smokes">
        <ComboboxInput
          data-testid="input"
          name="awesome"
          onChange={handleChange}
        />
        {results && (
          <ComboboxPopover portal={false}>
            <p>
              <button>Test focus</button>
            </p>
            <ComboboxList>
              {results.slice(0, 10).map((result, index) => (
                <ComboboxOption
                  key={index}
                  value={`${result.city}, ${result.state}`}
                />
              ))}
            </ComboboxList>
          </ComboboxPopover>
        )}
      </Combobox>
    </div>
  );
}

function useCityMatch(term: string) {
  let throttledTerm = useThrottle(term, 100);
  return useMemo(
    () =>
      term.trim() === ""
        ? null
        : matchSorter(cities, term, {
            keys: [item => `${item.city}, ${item.state}`],
          }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [throttledTerm]
  );
}
