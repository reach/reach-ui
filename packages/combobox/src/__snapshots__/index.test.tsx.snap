// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Combobox /> should match the snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Clientside Search
    </h2>
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="listbox--holy-smokes"
      data-reach-combobox=""
      id="holy-smokes"
      role="combobox"
    >
      <input
        aria-autocomplete="both"
        data-reach-combobox-input=""
        data-testid="input"
        id="listbox--holy-smokes"
        name="awesome"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Combobox /> should match the snapshot 2`] = `
<DocumentFragment>
  <div>
    <h2>
      Clientside Search
    </h2>
    <div
      aria-expanded="true"
      aria-haspopup="listbox"
      aria-owns="listbox--holy-smokes"
      data-reach-combobox=""
      id="holy-smokes"
      role="combobox"
    >
      <input
        aria-autocomplete="both"
        data-reach-combobox-input=""
        data-testid="input"
        id="listbox--holy-smokes"
        name="awesome"
        value="e"
      />
    </div>
  </div>
</DocumentFragment>
`;
